R/zchunk_LA100.IEA_downscale_ctry.R

```diff
@@ -0,0 +1,308 @@
+    IEA_product_downscaling <- get_data(all_data, "energy/mappings/IEA_product_downscaling")
+    IEA_ctry <- get_data(all_data, "energy/mappings/IEA_ctry")
+    L100.IEA_en_bal_ctry_hist_preprocessed <- get_data(all_data, "energy/L100.IEA_en_bal_ctry_hist")
+
+    # If the (proprietary) raw IEA datasets are available, go through the full computations below
+    # If not, use the pre-saved summary file (i.e., the output of this chunk!) assuming it's available
+    if(!is.na(en_nonOECD) && !is.na(en_OECD)) {
+
+      # Full calculation from raw data
+      # The two IEA datasets are LARGE and we keep them in wide format for most of the computations
+      # below, as they are very expensive to reshape. As a result, much of the logic below closely
+      # parallels that in the original data system, and there are fewer dplyr-style pipelines.
+
+      # Subset only the relevant years and combine OECD with non-OECD
+      hy <- intersect(intersect(HISTORICAL_YEARS, colnames(en_OECD)), colnames(en_nonOECD))
+      cols <- c("COUNTRY", "FLOW", "PRODUCT", hy)
+      bind_rows(en_OECD[cols], en_nonOECD[cols]) %>%
+        # rename fuels with inconsistent naming between the two databases
+        mutate(PRODUCT = replace(PRODUCT,
+                                 PRODUCT %in% c("Natural Gas", "Other Kerosene", "Total"),
+                                 c("Natural gas", "Other kerosene", "Total of all energy sources"))) ->
+        L100.IEAfull
+
+      # UP FRONT ADJUSTMENTS (UFA) original lines 42-67
+
+      # UFA1. Nearly the entire supply of natural gas in other Africa between 2001 and 2004 is allocated
+      # to GTL plants operating at nearly 100% efficiency. We adjust the energy input quantities
+      # to avoid negative values later on.
+      GTL_COEF <- 1.7
+      GTL_ADJ_YEARS <- as.character(intersect(2001:2004, HISTORICAL_YEARS))
+      if(length(GTL_ADJ_YEARS)) {
+        GTL_gas_entries <- with(L100.IEAfull, COUNTRY == "Other Africa" & FLOW == "TGTL" & PRODUCT == "Natural gas")
+        GTL_oh_entries <- with(L100.IEAfull, COUNTRY == "Other Africa" & FLOW == "TGTL" & PRODUCT == "Other hydrocarbons")
+        L100.IEAfull[GTL_gas_entries, GTL_ADJ_YEARS] <- L100.IEAfull[GTL_oh_entries, GTL_ADJ_YEARS] *
+          GTL_COEF * -1     # Multiply by -1 because other hydrocarbons are the output and have a different sign
+      }
+
+      # UFA2. South Africa has a coal-to-gas IO coef in the gas works sector of about 5:1, and low natural
+      # gas consumption in other sectors. The coal inputs are overridden here as the gas output times an
+      # exogenous IO coef
+      COAL_TO_GAS_COEF <- 1.3
+      CTG_entries <- L100.IEAfull$COUNTRY == "South Africa" & L100.IEAfull$FLOW == "TGASWKS"
+      # Only use other bituminous coal; no need to maintain distinction between coal (if no detail) and other bituminous coal
+      L100.IEAfull[CTG_entries & L100.IEAfull$PRODUCT == "Hard coal (if no detail)", hy] <- 0
```

Comment:
Ha, didn't mean to imply that! This particular row can be removed b/c it's with categories (sub-bituminous coal, other bituminous coal) that are redundant. We CAN absolutely remove zeroes earlier in the processing, but without further downstream checks and modifications, it might wreck later on due to missing/null categories.